docs/proposals/20200506-conditions.md

```diff
@@ -0,0 +1,363 @@
+
+This proposal aims to be consistent with the target state of conditions in Kubernetes (see [KEP](https://github.com/kubernetes/enhancements/pull/1624)).
+
+At the same time, we should think that Cluster API presents some specific challenges that are not common
+to the core Kubernetes objects:
+
+- Cluster API has a high number of long-running operations, and this makes it crucial for the success of
+  this proposal to introduce a clean way to provide evidence about what is happening during those operations,
+  and not only to show the final state of the operations.
+- Cluster API has a complex hierarchy of objects, and the informative value of a condition, especially in
+  the core types, is highly correlated to the possibility to collect and summarize meaningful information
+  from the underlying hierarchy of objects.
+- Cluster API defines a set of core objects and extension points (e.g. bootstrap, control plane and infrastructure).
+  Extension points are defined by a set of contracts, which each provider adheres to. To provide a great user
+  experience, core and provider contracts should be extended to allow for better coordination and visibility
+  based on conditions.
+
+#### Data Model Changes
+
+In this section we outline the proposed API changes.
+
+All the changes can be implemented in the v0.3.x timeframe (no breaking changes, only additions).
+
+Following types should be defined in Cluster API / v1alpha3 api.
+
+```golang
+// ConditionType is a valid value for Condition.Type.
+type ConditionType string
+
+// ConditionSeverity expresses the severity of a Condition Type failing.
+type ConditionSeverity string
+
+const (
+  // ConditionSeverityError specifies that a failure of a condition type
+  // should be viewed as an error.
+  ConditionSeverityError ConditionSeverity = "Error"
+
+  // ConditionSeverityWarning specifies that a failure of a condition type
+  // should be viewed as a warning, but that things could still work.
+  ConditionSeverityWarning ConditionSeverity = "Warning"
+
+  // ConditionSeverityInfo specifies that a failure of a condition type
+  // should be viewed as purely informational, and that things could still work.
+  ConditionSeverityInfo ConditionSeverity = "Info"
+
+  // ConditionSeverityNone should apply only if the condition is in state "True".
+  // As ConditionSeverityNone is the default for conditions we use the empty string (coupled with omitempty)
+  ConditionSeverityNone ConditionSeverity = ""
+)
+
+// Conditions define a readiness condition for a Cluster API resource.
+type Condition struct {
+   // Type of condition.
+   // +required
+   Type ConditionType `json:"type" description:"type of status condition"`
+
+   // Status of the condition, one of True, False, Unknown.
+   // +required
+   Status corev1.ConditionStatus `json:"status"`
+
+   // Severity with which to treat failures of this type of condition.
+   // When this is not specified, it defaults to Error.
```

Comment:
Why not make it a required field? Defaulting to `Error` might result in unexpected consequences

Reply:
Severity is relevant only for failed conditions